packages/edit-site/src/components/secondary-sidebar/list-view-sidebar.js

```diff
@@ -28,12 +35,29 @@ export default function ListViewSidebar() {
 	} );
 	const { selectBlock } = useDispatch( blockEditorStore );
 	const { setIsListViewOpened } = useDispatch( editSiteStore );
+
+	const constrainedTabbingRef = useConstrainedTabbing();
 	const focusOnMountRef = useFocusOnMount( 'firstElement' );
+	const focusReturnRef = useFocusReturn();
```

Comment:
Can't you just replace all these three hooks with useDialog instead?

Reply:
Not quite: `useDialog` also closes on click outside, which kinda defeats the "persistent" in "persistent list view". 😅 

(FYI, I did try it in 68606ba6372c6e12fae144073bf8b2e138404ae9, before noticing the click-outside behaviour)

Reply:
> Not quite: useDialog also closes on click outside

As I shared on the issue, this is the exact problem I tried to hightlight. I'm not sure removing that behavior is a good thing in terms of a11y.

Reply:
In other words, IMO, a persistent sidebar for me need a "checkbox"/"toggle" somewhere that says "show/hide the sidebar" but it shouldn't have a button that "opens the sidebar and moves focus to the sidebar" as well. It should have more a "setting" rather than an button that doesn't an "action" (open and navigate) as otherwise moving out would logically mean "close and navigate back"

Reply:
Here's a concrete problem of the current implementation:

 - Open the sidebar
 - Click outside
 - Click "Tab" repeatedly until you reach the sidebar, you'll get stuck in the sidebar because of the "constrained tabbing" behavior (which should only apply to dialogs) which means we created a "tab trap" which violates the WCAG  rules.

Reply:
TBH I think the current behaviour works well enough. 🤔 
The List View is openable via keyboard in 2 different ways (tabbing to the toggle button, or via shortcut), closable in 2 ways as well (ESC, or via shortcut), and can move focus to the canvas by selecting a list item.

This said, this stuff is very particular and I don't know all the context and discussions happened for the other sidebars, so I'm happy to disagree and update anyway.
Although, I've been working following designers mockups, and they might be more familiar with the established sidebar interactions? cc @jameskoster

Reply:
I know that's what the designers want, I'm just trying to illustrated that there's bug in the proposed behavior :). 

The only alternative I can think are:

Make the button in the top bar behave as a "setting" and not as an action. Which means don't focus the sidebar on mount. (Might make navigating to the sidebar a bit harder for screen readers users as it'd require a shortcut or using the region navigation but at least it keeps the keyboard navigation sane)

The other would would be to move the button to open the sidebar to the actual sidebar (top of it), you know just like the toggles we see in websites sometimes to show left menus. This means there's no need for focus management since the sidebar is shown right after the button. I'm sure designers can find a way to make this look good  :P  

Maybe there are other alternatives, but if we want it to be a persistent sidebar, we shouldn't be trying to make it work like a dialog (focus on mount and constrained tabbing), these things only work if you have (close on focus leave).

Reply:
> Make the button in the top bar behave as a "setting" and not as an action.

It's funny you mention that, because I happen to think this direction could be the way to go in the long term for various UI components. Toggles for things like list view and breadcrumb could potentially be added to the "View" menu (related: https://github.com/WordPress/gutenberg/pull/29151). Subsequently the existing toggles for top toolbar, spotlight mode, full-screen mode, etc, could conceivably join them there as well.

Rough mockup for this context:

![block list](https://user-images.githubusercontent.com/846565/109196293-6c8cd980-7793-11eb-8365-b984f2b50822.gif)

> The other would would be to move the button to open the sidebar to the actual sidebar

I'm struggling to picture how this would work. If the button is inside the sidebar, and the sidebar is hidden, how do you click the button? :D

Reply:
@youknowriad apologies, I missed the [comment](https://github.com/WordPress/gutenberg/pull/28637#discussion_r582988035) where you outlined the issue! Must have popped up while I was writing my answer.

I didn't notice how the constrained tabbing affected the sidebar, but you are totally right about it.
I've removed the constrained tabbing leaving only these behaviours:
- focus on mount (selected element if any, or first element)
- close on ESC
- focus return

The tab is not trapped anymore, although the keyboard navigation is still confusing.
The `BlockNavigation` component renders a tree, which is navigable with the arrows instead of tab.
If you are focusing a list item, tab moves you to the canvas, selecting a block, which in turns highlights the corresponding list item. At that point it's very hard to understand if you are focusing the canvas or the list.

Though, this confusion happens even when clicking around with the mouse, as the selected list item is very heavy and in-your-face, compared to the selected block outline.
We have discussed about it for a while, and decided to move it to a follow up dedicated to the style of the sidebar.

I think that if you are cool with the fact that the tab is not trapped anymore, we can move forward with this for both the style and a decision about how to appropriately toggle the sidebar.

IMHO this PR is really valuable in itself even with its rough edges, and it's limited to the Site Editor so it won't cause nasty regressions in production features.
In other words, I'm keen to merge this (provided the code is good, of course), and iterate to fine tune the UX.

Reply:
> The other would would be to move the button to open the sidebar to the actual sidebar

> > I'm struggling to picture how this would work. If the button is inside the sidebar, and the sidebar is hidden, how do you click the button? :D

I think this idea is that the toggle button should be above the sidebar (similar to where the list view button currently exists outside of this PR) and more 'connected' to it.  Since it would be a 'setting' and not directing focus into the sidebar, it would be important for the next 'tab' from the toggle button to lead into the sidebar if it is open.

Reply:
> I'm struggling to picture how this would work. If the button is inside the sidebar, and the sidebar is hidden, how do you click the button? :D

I was suggesting here that when it's "closed", it's not "entirely closed" but instead the "button" is the only thing in the sidebar (and the sidebar doesn't have any background or borders, so we only see the button)

That said, moving to "View" is actually very good and is what makes the most sense to me. 

Now, about the current behavior (after dropping the constrained tabbing), I think I'm ok with it personally, though it's arguable whether other screen reader users might think.